### ISSUE TITLE
Format on save deletes text

### ISSUE THREAD

**Issue by samrickman** [quarto-dev/quarto#246]
Frequently, I am writing a long `qmd` file, and when I save (which auto formats), lines get added to or deleted from R code chunks. These line breaks can be in the middle of words, and the deleted lines often contain code.

Weirdly, this does not happen every time I save, but once I've been editing a document for around 20 mins or more, it is practically guaranteed to start happening. It is quite irritating if a code block somewhere far away is changed when you save. I often don't notice until I try to render, at which point some lines of code are missing, the file won't render and I have saved over the working version.

Here is a gif of what I am talking about. I thought this might be caused by another extension so here I am manually using `Quarto: Format Cell` in the command palette. However, this also happens when I save (unless I use `File: Save without Formatting`):

![quarto-save-gif](https://github.com/quarto-dev/quarto/assets/46163241/a7768611-1b2c-4fcb-8dbe-97d29ff9edd8)

You can see by the third format line breaks have been created which cause syntax errors. Although it does not happen here, lines are also often deleted.

Here is the full text of the document in the gif:

`````
---
title: "Discussion"
format: pdf
---

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean eget rutrum eros. Nullam lacinia efficitur risus, at sagittis metus fringilla sed. Sed eget vulputate ipsum. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Integer pharetra iaculis erat, eu placerat massa varius eget. Etiam rutrum quis sem sed rhoncus. Ut a massa eget mauris hendrerit tincidunt in id libero. In mollis, tellus a euismod venenatis,  metus orci maximus nulla, quis tristique quam tortor eget massa.

```{r logreg, echo = FALSE}
#| label: tbl-logreg
#| tbl-cap: "Logistic regression: factors in structured data associated with loneliness and social isolation"

library(data.table)
library(kableExtra)
logreg <- fread("./csv/log_reg_wide.csv")

setnames(logreg, "explanatory", " ")

logreg |>
    kbl(
booktabs = T) |>
    kable_styling(latex_options = c("HOLD_position", "scale_down")) |>
    add_header_above(
        c(
            " " = 1, "Type A" = 2, "Type B" = 4
        )
    ) |>    footnote("*** < 0.001; ** <0.01; * <0.05; . <0.1", general_title = "")
```

Sometimes text disappears when I save ad a

Fusce mattis efficitur consectetur. Sed dictum conadgue finibus. Curabitur tristique eleifend arcu, eu laoreet nisi placerat eget. Nullam tincidunt diam vulputate, hendrerit risus ut, fringilla lorem. Nullam vel ipsum nisi. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vestibulum at dictum diam, quis sodales elit. Duis eget quam ut nisi tempor consectetur.


Suspendisse in mattis nisl. Fusce ut interdum erat. Pellentesque consequat purus vel suscipit vehicula. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque purus lectus, tempus in aliquam ultrices, pretium vitae ligula. Nunc metus neque, maximus quis nibh a, condimentum consequat purus. Pellentesque lacus mi, aliquam et faucibus vel, maximus ac magna. Aliquam non tellus maximus, convallis nisi vitae, ultricies magna. Sed non ligula sit amet orci faucibus ullamcorper. Curabitur sagittis quam et eros aliquam tempor. Aliquam laoreet varius pharetra. Maecenas non lacinia lectus.
`````

### Quarto
Quarto VScode extension version: v1.94.0

### VScode
Version: 1.81.1 (user setup)
Commit: 6c3e3dba23e8fadc360aed75ce363ba185c49794
Date: 2023-08-09T22:22:42.175Z (1 wk ago)
Electron: 22.3.18
ElectronBuildId: 22689846
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.19045

### R
R version 4.1.3 (2022-03-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)  
Running under: Windows 10 x64 (build 19045)

I can provide my entire `settings.json` if helpful - it's quite long. The only parts which mention `quarto` or R are:

```
  "[quarto]": {
    "cSpell.showAutocompleteSuggestions": true,
    "mathjax.theme": "dark"
  },
  "cSpell.enableFiletypes": [
    "quarto",
    "qmd"
  ,
  // R stuff
  "r.rterm.windows": "C:\\Users\\rickman\\AppData\\Local\\Programs\\Python\\Python38\\Scripts\\radian.exe",
  "r.rterm.option": [],
  "r.sessionWatcher": true,
  "files.associations": {
    "*.rmd": "markdown"
  },
  "r.bracketedPaste": true,
```

I imagine the source of this bug will be very difficult or impossible to track down if it can't be reproduced. Let me know if there's any more info that I can provide.

**Comment by jjallaire** [comment]
Thanks for the report and the repro! Even if I can't reproduce locally the fact that this can happen at all tells me that some assumption we are making doesn't hold. It may be that we need to take a more conservative approach to formatting (currently we send the entire document with commented lines in the non-cell regions, we may instead need create a brand new document with only the code cell (then offset the edits as required).

**Comment by jjallaire** [comment]
I think that this issue might be similar to what was reported here: https://github.com/quarto-dev/quarto/issues/190

What's likely happening is that the temporary R file we create to do formatting (`.intellisense.r`) is getting formatted by the R extension, which throws off all of the line and character counts we use when applying the changes. There were two solutions in this case:

1) Disable format on save in the R extension (this though is of course overly punitive and not a "real" solution). It would nevertheless interesting to know whether this resolves the problem in your case.

2) We inject a comment in `.intellisense.r` to turn off formatting. We did this for `styler` here (https://github.com/quarto-dev/quarto/commit/ec897f43d3487856676bfed17a56573872ad31ea#diff-553a848eeee5588204aa57192c9ba1a2f43ffe3f97e48d5b20f639e119255ddeL52-L56), however perhaps you are using a different formatter that we need add another comment to disable?

**Comment by samrickman** [comment]
@jjallaire yes! I was actually going to suggest it had something to do with that issue, as I've noticed the issue only seems to happen when `.intellisense.r` appears as an open tab (which you can see is the case in the gif). It stops for a while when I close that tab (which makes sense as why would the formatter format files which aren't open).  But then at some point a new `.intellisense.r` tab opens up.

I don't know what causes the tab to open: it's not always there when I am editing and I don't manually open it.

I am just using the default R formatter from the _R Extension for Visual Studio Code_, and no others, for `.r` files. Next time the issue arises I will temporarily disable it and let you know if it changes anything.

**Comment by jjallaire** [comment]
Okay, I think the issue then is that because you have a larger document the formatting takes a few seconds and this allows `.intellisense.r` to open (whereas normally it never opens). Once its open then format-on-save is basically happening twice!

We can probably work around this by (a) Refusing to apply formatting if we notice the document has been mutated; or (b) Trying to detect `.intellisense.r` being open using the tabs api (https://github.com/microsoft/vscode/commit/aa69f3d7623c464aba726d12ea0d83428f43e8b9#diff-0a75aed19c118603eb96332bc0b9c2d7867f4182346d16d18b7fc31b6ceeb321) and closing it. 

In either case we might not actually do the formatting (but at least we wouldn't lose edits!).

I'm a bit surprised that the `# styler: off` comment we inject at the top isn't resulting in the R extension standing down on formatting.

**Comment by samrickman** [comment]
If that is the cause of issue then what about adding to `settings.json`:

```
  "files.associations": {
    ".intellisense.r": "plaintext"
  },
```

I just did this, created a 50k line `qmd` file (for testing only!) and formatted a code cell with the quarto extension. A few things:

1. As expected, when I open `.intellisense.r`, it has `# styler: off` at the top, a lot of blank lines where the prose was, and the code chunk on the correct lines.
2. VScode now sees `.intellisense.r` as plain text - there's no longer any syntax highlighting, nothing is formatted on save and if I try to manually format it says there's no installed formatter for plain text. 
3. The formatting worked this time - no lines were created or deleted. 

I've never created a vscode extension so if this is a good approach I'm not sure exactly how an extension can edit `settings.json` (or if it needs to directly) - but might this be a way forward? I don't know if `.intellisense.r` is used for anything else that would need it to be seen as a `.r` file.

**Comment by samrickman** [comment]
@jjallaire this seems to have fixed the issue for me. Not sure if this is helpful but once I looked into how to include this setting in a package it turned out it was just a question of putting it in the right place in `package.json`, so I created this [PR](https://github.com/quarto-dev/quarto/pull/249) which does that.

I don't submit PRs very often - I've tried to merge with main but if that's not the right branch let me know. Alternatively if you want to go in another direction let me know and I can close the PR.

**Comment by jjallaire** [comment]
@samrickman Thanks so much for the additional investigation and even finding the solution! Will review and merge the PR shortly. Will also let you know when an updated version of the extension with the fix is published.

**Comment by samrickman** [comment]
@jjallaire thank you for your quick responses and merging the PR so speedily - every step of this issue has been a dream. And also thanks for creating quarto (and rmd before it)! They're amazing tools and I can't imagine how anyone managed without them.

**Comment by jjallaire** [comment]
Thanks for the kind words!

**Comment by jjallaire** [comment]
Unfortunately this fix ended up disabling autocomplete! (see https://github.com/quarto-dev/quarto/issues/253) so I had to revert it. I will need to see if I can find another way (it might just be giving up on format on save if the document changes from underneath us -- note that this might not even be enough if we are first applying our formats and _then_ the R extension is applying its formats).

**Comment by samrickman** [comment]
Oh dear that's disappointing - but I was half-expecting some sort of unintended consequence. At least you found it early. It's a public holiday in the UK and I'm away until mid next week but I'll have a think about it and see if I can come up with a different approach on my return. 

As you say essentially the double formatting is the problem - particularly because the formatting rules don't seem to agree with the R extension. Which bit of the repo are the formatting rules in? Could they be aligned with whatever the user has set up? Obviously that would need to be across all languages...

**Comment by jjallaire** [comment]
For the sake of safety I have for now disabled format document for R (which also disables format on save).

The key is going to be convincing the R extension to NOT do a format on save for the `.intellesense.r` file. I assumed that `# styler: off` at the top would be enough to do this but it seems not. If there is another incantation in the form of a comment in the file then that should also take care of the problem.

**Comment by samrickman** [comment]
I have been investigating a bit more. I do not think the problem is `styler` running when it isn't supposed to. Further than that, I'm afraid I have more questions than answers.  I rolled back to  `v1.94.0` to see what was happening. In particular I wanted to check two things:

1. Why should re-formatting the already formatted R script change the lines? In principle once a file has been through a formatter once, another pass through the same formatter should not change anything.
2. Does this only happen with R?

With respect to the first question, I noticed that my `.qmd` files have CRLF line endings, but my `.intellisense.r` file has LF line endings. It looks like `styler` [only supports LF](https://github.com/r-lib/styler/blob/1976817c5a894b86c313f1c30ee9cf381ed787c6/R/parse.R#L24). However, I think this may be a red herring, as `styler` should not be running given the `# styler: off` comment. 

I tried manually running `styler` on a copy of the `.intellisense.r` file with `styler::style_file()` and it did not make any changes. Similarly, I created a deliberately badly-formatted `.r` file prepended with `# styler: off` and tried styling it with both Format on Save in VS Code and also `styler::style_file()` from R. In both cases it acted as intended and did not make any changes. This makes there may be something else going on that's not caused by `styler`. This is borne out by what happened with a Python `.qmd` file.

## Odd Python behaviour

With respect to whether this only happens with R, I have encountered some really weird behaviour. I copied and pasted the example Python file from the Quarto [docs](https://quarto.org/docs/computations/python.html) until it was about 5k lines. When I save it, it seems to overwrite the entire `.qmd` file with the `.intellisense.py` file:

![quartopython](https://github.com/quarto-dev/quarto/assets/46163241/ec265e6e-089d-4e8c-980b-386ad6362003)

This was with with `v1.94.0`. With the current version I get the same behaviour except the syntax highlighting has changed (it's now highlighted like a `.py` file rather than a `.qmd` file:

![quartopython2](https://github.com/quarto-dev/quarto/assets/46163241/29357c78-c96a-4752-af54-c8c5957bf768)

In a way this is reassuring. I can believe that some other people occasionally have lines deleted and haven't reported it but it seems unlikely that lots of people are having the entire document overwritten - so this may all just be something very strange with my environment. I am using a pretty old laptop (6th gen i7 CPU) so it may take longer than most to save.

But I am really stumped by this Python behaviour which I haven't come across before. It does not happen with normal length Python `.qmd` files - they have to be around 3.1k lines before it happens. I think your theory is right that when it takes particularly long to save a file, the `.intellisense.*` file is open for long enough for weird things to happen. 

I am not convinced the issue is double-styling being applied, given the R `styler` package seems to work as intended with `# styler: off` set, and the odd stuff happening in Python with very long files. But if it is not then I don't know what it might be. Do you have any ideas?

**Comment by jjallaire** [comment]
Thanks again for all of this additional context! I am convinced that our whole strategy for applying formatting is overly ambitious (especially given that race conditions crop up for both R and Python with longer documents). In version 1.96.0 (now available) I have switched to a much more conservative strategy: we only format the currently active cell (rather than attempting to format all of the cells in the document). This provides both better performance and less exposure to race conditions. I think it is also fairly sane / understandable that formatting the "document" means formatting the cell currently being edited. Note that with this change format on save is re-enabled for R (but only the current cell is formatted).

**Comment by m-clark** [comment]
@jjallaire Thought I'd note in case it helps with development in other ways.  I was having very similar formatting issues, but in troubleshooting I found that it may only be happening in cells with chunk options at the top. Versions (Quarto VSCode Extension 1.97, Quarto 1.4.339)

Starting point: 

<img width="834" alt="image" src="https://github.com/quarto-dev/quarto/assets/4720569/ed33c333-5a08-4706-b10d-6631a1fb5544">


On save (this is fine, though it'd be nice to control arrows :)):

<img width="823" alt="image" src="https://github.com/quarto-dev/quarto/assets/4720569/e0d5658b-87a8-4f35-8b2e-5b9785c78eba">


With cell options added to starting point and saving, auto format breaks. Arrows are not placed correctly, line returns are added, some lines are deleted, etc.:

<img width="829" alt="image" src="https://github.com/quarto-dev/quarto/assets/4720569/c094d56b-520d-4f7d-8b8a-2d2e5ba4f512">

**Comment by jjallaire** [comment]
@m-clark Thank you so much for your persistence and repro case! I did indeed find another issue related to the presence of YAML options which is now fixed here: https://github.com/quarto-dev/quarto/commit/cc3a7d46135e8d1a2a4d5e456156bb461c8c7612

This is available now in the v1.98.0 release.

**Comment by m-clark** [comment]
Thanks folks! The effort is much appreciated!
